// promise-stuff/dist/lib/operators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const static_operators_1 = require("./static-operators");
const common_1 = require("./common");
/**
 * A module that provides static bindings for functions implemented by [[ExtendedPromise]].
 */
var Operators;
(function (Operators) {
    /**
     * Returns a `StaticOperators` module, which provides static bindings for functions implemented by a [[ExtendedPromiseConstructor]].
     * The `StaticOperators` instance provides functions that use the constructor `ctor` internally.
     * @param {BasicPromiseConstructor<P extends BasicPromise<*>>} ctor The promise constructor for which to create an operators module.
     * @returns {StaticOperators<BasicPromise<*>>}
     * @constructor
     */
    function For(ctor) {
        return static_operators_1.StaticOperators.create(ctor);
    }
    Operators.For = For;
    /**
     * Static version of the `race` function.
     * @see [[ExtendedPromise.race]]
     */
    function race(promise, ...others) {
        let ctor = promise.constructor;
        others.unshift(promise);
        if ("race" in ctor) {
            return ctor.race(others);
        }
        return new ctor((resolve, reject) => {
            let done = false;
            for (let promise of others) {
                promise.then(x => {
                    !done && resolve(x);
                    done = true;
                }, err => {
                    !done && reject(err);
                    done = true;
                });
            }
        });
    }
    Operators.race = race;
    /**
     * Static version of the `and` function.
     * @see [[ExtendedPromise.and]]
     * @param {P} promise
     * @param {PromiseLike<T>} others
     * @returns {BasicPromise<*[]>} The return type of this function is meant to be `P<T[]>`, where `P` is the input promise type.
     */
    function and(promise, ...others) {
        let ctor = promise.constructor;
        others.unshift(promise);
        if ("all" in ctor) {
            return ctor.all(others);
        }
        return new ctor((resolve, reject) => {
            let arr = [];
            let count = 0;
            let done = false;
            for (let i = 0; i < others.length; i++) {
                let promise = others[i];
                promise.then(x => {
                    if (done)
                        return;
                    arr[i] = x;
                    count++;
                    if (count === others.length) {
                        resolve(arr);
                        done = true;
                    }
                }, err => {
                    reject(err);
                    done = true;
                });
            }
        });
    }
    Operators.and = and;
    /**
     * Static version of the `fallback` function.
     * @see [[ExtendedPromise.fallback]]
     * @param {P} promise
     * @param {PromiseLike<T> | ((reason: any) => (PromiseLike<T> | T)) | T} others
     * @returns {P}
     */
    function fallback(promise, ...others) {
        let ctor = promise.constructor;
        others.unshift(promise);
        let lastFailure = null;
        return new ctor((resolve, reject) => {
            let resolveOn = i => {
                if (i >= others.length)
                    return reject(lastFailure);
                let current = others[i];
                if (common_1.isFunction(current)) {
                    current = current(lastFailure);
                }
                if (common_1.isPromiseLike(current)) {
                    current.then(x => resolve(x), err => {
                        lastFailure = err;
                        if (i >= others.length - 1) {
                            return reject(err);
                        }
                        else {
                            return resolveOn(i + 1);
                        }
                    });
                }
                else {
                    return resolve(current);
                }
            };
            others.filter(other => common_1.isPromiseLike(other)).forEach(promise => promise.then(undefined, () => {
            }));
            resolveOn(0);
        });
    }
    Operators.fallback = fallback;
    /**
     * Static version of the `delay` function.
     * @see [[ExtendedPromise.delay]]
     * @param {P} promise
     * @param {number | Date} time
     * @returns {P}
     */
    function delay(promise, time) {
        let ctor = promise.constructor;
        let operators = For(ctor);
        return promise.then(x => operators.wait(time).then(() => x), x => operators.wait(time).then(() => {
            return common_1.makeRejectingPromise(ctor, x);
        }));
    }
    Operators.delay = delay;
    /**
     * Static version of the `each` function.
     * @see [[ExtendedPromise.each]]
     * @param {P} promise
     * @param {AsyncCallback<any, void>} action
     * @returns {P}
     */
    function each(promise, action) {
        return promise.then(x => {
            let p = action(x, true);
            if (p instanceof promise.constructor) {
                return p.then(() => x);
            }
            else {
                return x;
            }
        });
    }
    Operators.each = each;
    /**
     * Static version of the `lastly` function.
     * @see [[ExtendedPromise.lastly]]
     * @param {P} promise
     * @param {AsyncCallback<any, void>} action
     * @returns {P}
     */
    function lastly(promise, action) {
        let ctor = promise.constructor;
        return promise.then(x => {
            let p = action(x, true);
            if (p instanceof ctor) {
                return p.then(() => x);
            }
        }, err => {
            let p = action(err, false);
            if (p instanceof ctor) {
                return p.then(() => common_1.makeRejectingPromise(ctor, err));
            }
        });
    }
    Operators.lastly = lastly;
    /**
     * Static version of the `mustNot` function.
     * @see [[ExtendedPromise.mustNot]]
     * @param {P} promise
     * @param {AsyncCallback<T, any>} failReason
     * @returns {P}
     */
    function mustNot(promise, failReason) {
        let ctor = promise.constructor;
        return promise.then(result => {
            let p = failReason(result, true);
            if (common_1.isPromiseLike(p)) {
                return p.then(err => {
                    if (err) {
                        return common_1.makeRejectingPromise(ctor, err);
                    }
                    else {
                        return result;
                    }
                }, err => common_1.makeRejectingPromise(ctor, err));
            }
            else if (p) {
                return common_1.makeRejectingPromise(ctor, p);
            }
            else {
                return result;
            }
        });
    }
    Operators.mustNot = mustNot;
    /**
     * Static version of the `stall` function.
     * @see [[ExtendedPromise.stall]]
     * @param {P} promise
     * @param {number | Date} time
     * @returns {P}
     */
    function stall(promise, time) {
        let deadline = Date.now() + common_1.normalizeTime(time);
        let ctor = promise.constructor;
        return new ctor((resolve, reject) => {
            promise.then(x => {
                let now = Date.now();
                if (now >= deadline) {
                    resolve(x);
                }
                else {
                    setTimeout(() => {
                        resolve(x);
                    }, now - deadline);
                }
            }, err => {
                let now = Date.now();
                if (now >= deadline) {
                    reject(err);
                }
                else {
                    setTimeout(() => {
                        reject(err);
                    }, now - deadline);
                }
            });
        });
    }
    Operators.stall = stall;
    /**
     * Static version of the `timeout` function.
     * @see [[ExtendedPromise.timeout]]
     * @param {P} promise
     * @param {number | Date} time
     * @param {() => (PromiseLike<T> | T)} onTimeout
     * @returns {P}
     */
    function timeout(promise, time, onTimeout) {
        let ctor = promise.constructor;
        onTimeout = onTimeout || (() => Promise.reject(new Error("Timed Out")));
        let stat = For(ctor);
        let timeoutTk = {};
        return race(promise, stat.wait(time, timeoutTk)).then(x => {
            if (x === timeoutTk) {
                let result = onTimeout();
                return result;
            }
            return x;
        });
    }
    Operators.timeout = timeout;
    /**
     * Static version of the `test` function.
     * @see [[ExtendedPromise.test]]
     * @param {P} promise
     * @returns {BasicPromise<boolean>} The return type is meant to be `P<boolean>`, where `P` is the promise type.
     */
    function test(promise) {
        return promise.then(x => true, x => false);
    }
    Operators.test = test;
    /**
     * Static version of the `invert` function.
     * @see [[ExtendedPromise.invert]]
     * @param {P} promise
     * @returns {BasicPromise<*>} The return type of this function is meant to be `P<any>`, where `P` is the promise type.
     */
    function invert(promise) {
        let ctor = promise.constructor;
        return promise.then(x => common_1.makeRejectingPromise(ctor, x), err => err);
    }
    Operators.invert = invert;
})(Operators = exports.Operators || (exports.Operators = {}));
