// wampus/wrappers/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* istanbul ignore file */
const types_1 = require("../core/errors/types");
/**@internal*/
var Errors;
(function (Errors) {
    function unknownEvent(name) {
        return new Error(`Unknown event name ${name}.`);
    }
    Errors.unknownEvent = unknownEvent;
    let Creation;
    (function (Creation) {
        function unknownSerializerName(object) {
            return new types_1.WampusInvalidArgument("Unknown serializer name '{name}'.", {
                name: object
            });
        }
        Creation.unknownSerializerName = unknownSerializerName;
        function unknownSerializer(object) {
            return new types_1.WampusInvalidArgument("Given invalid serializer.", {});
        }
        Creation.unknownSerializer = unknownSerializer;
        function unknownTransportName(object) {
            return new types_1.WampusInvalidArgument("Unknown transport name '{name}'.", {
                name: object
            });
        }
        Creation.unknownTransportName = unknownTransportName;
        function unknownTransport(object) {
            return new types_1.WampusInvalidArgument("Given invalid transport.", {});
        }
        Creation.unknownTransport = unknownTransport;
    })(Creation = Errors.Creation || (Errors.Creation = {}));
})(Errors = exports.Errors || (exports.Errors = {}));
