// wampus/core/protocol/message.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The message type IDs for all messages in the WAMP protocol specification.
 * @see [Basic Profile]{@link https://wamp-proto.org/_static/wamp_latest.html#message-codes-and-direction}
 * @see [Advanced Profile]{@link https://wamp-proto.org/_static/wamp_latest.html#message-codes-and-direction-0}
 */
var WampType;
(function (WampType) {
    WampType[WampType["HELLO"] = 1] = "HELLO";
    WampType[WampType["WELCOME"] = 2] = "WELCOME";
    WampType[WampType["ABORT"] = 3] = "ABORT";
    WampType[WampType["GOODBYE"] = 6] = "GOODBYE";
    WampType[WampType["ERROR"] = 8] = "ERROR";
    WampType[WampType["PUBLISH"] = 16] = "PUBLISH";
    WampType[WampType["PUBLISHED"] = 17] = "PUBLISHED";
    WampType[WampType["SUBSCRIBE"] = 32] = "SUBSCRIBE";
    WampType[WampType["SUBSCRIBED"] = 33] = "SUBSCRIBED";
    WampType[WampType["UNSUBSCRIBE"] = 34] = "UNSUBSCRIBE";
    WampType[WampType["UNSUBSCRIBED"] = 35] = "UNSUBSCRIBED";
    WampType[WampType["EVENT"] = 36] = "EVENT";
    WampType[WampType["CALL"] = 48] = "CALL";
    WampType[WampType["RESULT"] = 50] = "RESULT";
    WampType[WampType["REGISTER"] = 64] = "REGISTER";
    WampType[WampType["REGISTERED"] = 65] = "REGISTERED";
    WampType[WampType["UNREGISTER"] = 66] = "UNREGISTER";
    WampType[WampType["UNREGISTERED"] = 67] = "UNREGISTERED";
    WampType[WampType["INVOCATION"] = 68] = "INVOCATION";
    WampType[WampType["YIELD"] = 70] = "YIELD";
    WampType[WampType["CANCEL"] = 49] = "CANCEL";
    WampType[WampType["CHALLENGE"] = 4] = "CHALLENGE";
    WampType[WampType["AUTHENTICATE"] = 5] = "AUTHENTICATE";
    WampType[WampType["INTERRUPT"] = 69] = "INTERRUPT";
    WampType[WampType["_Unknown"] = -1] = "_Unknown";
})(WampType = exports.WampType || (exports.WampType = {}));
