// wampus/core/protocol/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * For shared registrations, determines the policy used to pick which registrant gets invoked when a procedure is called.
 */
var InvocationPolicy;
(function (InvocationPolicy) {
    /**
     * Single registration policy.
     */
    InvocationPolicy["Single"] = "single";
    /**
     *  The callees will be notified using a round-robin policy.
     */
    InvocationPolicy["RoundRobin"] = "roundrobin";
    /**
     * A callee will be notified randomly.
     */
    InvocationPolicy["Random"] = "random";
    /**
     * The first callee to register the procedure will handle it.
     */
    InvocationPolicy["First"] = "first";
    /**
     * The last callee to register a procedure will handle it.
     */
    InvocationPolicy["Last"] = "last";
})(InvocationPolicy = exports.InvocationPolicy || (exports.InvocationPolicy = {}));
/**
 * For pattern-based subscription and registration, the matching policy used to determine if a given event or procedure name matches the pattern.
 */
var MatchingPolicy;
(function (MatchingPolicy) {
    /**
     * Uses the prefix matching policy.
     */
    MatchingPolicy["Prefix"] = "prefix";
    /**
     * Uses the wildcard matching policy.
     */
    MatchingPolicy["Wildcard"] = "wildcard";
})(MatchingPolicy = exports.MatchingPolicy || (exports.MatchingPolicy = {}));
/**
 * Tells the router how to cancel a call.
 * @see [WAMP Spec - Cancellation]{@link https://wamp-proto.org/_static/wamp_latest.html#feature-definition-2}
 */
var CancelMode;
(function (CancelMode) {
    /**
     * The router will send an error to the caller, but won't send any message to the callee. The callee's response will be discarded when received.
     */
    CancelMode["Skip"] = "skip";
    /**
     * The WAMP router will send a message to the callee and send the callee's response to the caller.
     */
    CancelMode["Kill"] = "kill";
    /**
     * The WAMP router will send an error to the caller and a cancellation message to the callee. If the callee responds, its response is discarded.
     */
    CancelMode["KillNoWait"] = "killnowait";
})(CancelMode = exports.CancelMode || (exports.CancelMode = {}));
